client/components/data/sync-active-theme/index.tsx

```diff
@@ -0,0 +1,40 @@
+import { useEffect, useState } from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+import { waitFor } from 'calypso/my-sites/marketplace/util';
+import { requestActiveTheme } from 'calypso/state/themes/actions';
+import { isRequestingActiveTheme, getActiveTheme } from 'calypso/state/themes/selectors';
+
+interface SyncActiveThemeProps {
+	siteId: number;
+	themeId: string;
+	onThemeActive: CallableFunction;
+	maxAttempts?: number;
+}
+
+const SyncActiveTheme = ( {
+	siteId,
+	themeId,
+	onThemeActive,
+	maxAttempts = 20,
+}: SyncActiveThemeProps ) => {
+	const dispatch = useDispatch();
+	const requestStarted = useSelector( ( state ) => isRequestingActiveTheme( state, siteId ) );
+
+	const activeTheme = useSelector( ( state ) => getActiveTheme( state, siteId ) );
+
+	const [ attempts, setAttempts ] = useState( 0 );
+
+	useEffect( () => {
+		if ( activeTheme === themeId ) {
+			return onThemeActive();
+		}
+		if ( ! requestStarted && attempts < maxAttempts ) {
```

Comment:
Can we add a debug message or a Tracks event for the case where we hit/exceed maxAttempts? It feels like it would be _really_ useful to know how often that occurs, and what the total time was for that case.